docs/docs/architecture/client_broker_protocol.md

```diff
@@ -0,0 +1,1223 @@
+fields in this type:
+
+- `uri`: A string which contains the full BlazingMQ queue
+  [`URI`](https://github.com/bloomberg/blazingmq-sdk-java/blob/main/bmq-sdk/src/main/java/com/bloomberg/bmq/Uri.java)
+  to which client wants to attach.
+
+- `qId`: An integer identifier that the SDK and broker will use to identify any
+  message for this queue.  This integer needs to be unique only between this
+  SDK client and the broker.  A simple implementation would be to start with
+  zero, and simply increment the value any time the client opens a new queue.
+  The goal is to avoid passing the full queue URI in every message, and instead
+  just use this 4-byte integer to save space on the wire.
+
+- `subQueueIdInfo`: This field's type is
+  [`SubQueueIdInfo`](https://github.com/bloomberg/blazingmq/blob/9b692fe25f74543e954a27e30b6b15b0ae057c8d/src/groups/bmq/bmqp/bmqp_ctrlmsg.xsd#L364-L380).
+  This type captures details about the sub-stream of the queue.  **TODO**
+
+- `flags`: An integer which indicates the intent with which a BlazingMQ client
+  wants to open the queue.  Valid values can be found
+  [here](https://github.com/bloomberg/blazingmq/blob/9b692fe25f74543e954a27e30b6b15b0ae057c8d/src/groups/bmq/bmqt/bmqt_queueflags.h#L46).  Note that the `e_ADMIN` is reserved and
+  should not be used.
+
+- `readCount`: An integer whose value should be one if the client wants to
+  attach to the queue as a consumer, and zero otherwise.
+
+- `writeCount`: An integer whose value should be one if the client wants to
+  attach to the queue as a producer, and zero otherwise.
+
+- `adminCount`: An integer whose value should always be zero.
+
+[Here](https://github.com/bloomberg/blazingmq-sdk-java/blob/ed0eb848f5ac2897cfe32cc481575d689a4cb1c2/bmq-sdk/src/test/java/com/bloomberg/bmq/it/PlainProducerIT.java#L162-L203), we
+see the `PlainProducerIT.java` creating and sending an `OpenQueue` request.
+
+Upon success, the BlazingMQ broker responds to the `OpenQueue` request by
+sending an
+[`OpenQueueResponse`](https://github.com/bloomberg/blazingmq/blob/9b692fe25f74543e954a27e30b6b15b0ae057c8d/src/groups/bmq/bmqp/bmqp_ctrlmsg.xsd#L191-L211).  Let's go over the three
+fields in this type:
+
+- `originalRequest`: A field whose type is `OpenQueue` request, and it simply
+  carries the `OpenQueue` request sent by the client.
+
+- `routingConfiguration`: An integer which indicates information about message
+  routing strategy of the queue, as indicated by the broker.  Possible values
+  for the flag can be found
+  [here](https://github.com/bloomberg/blazingmq/blob/9b692fe25f74543e954a27e30b6b15b0ae057c8d/src/groups/bmq/bmqp/bmqp_routingconfigurationutils.h#L51-L61).
+
+- `deduplicationTimeMs`: An integer which indicates the time interval (in ms)
+  for which the SDK should keep the unacknowledged PUT messages in the
+  retransmission buffer before generating a local negative acknowledgement.
+  This field can be ignored in the initial implementation of the SDK.
```

Comment:
An example of what the C++/Java SDKs do (as in, beyond an initial implementation of the SDK), could be valuable, as in [this](https://github.com/bloomberg/blazingmq/blob/cce74cbe4f781995b3a9cb9d3f206f6b7f0a4452/src/groups/bmq/bmqimp/bmqimp_brokersession.cpp#L1196).